fuzz/fuzz_targets/roundtrip.rs

```diff
@@ -0,0 +1,71 @@
+#![no_main]
+
+use png::{FilterType, ColorType};
+#[macro_use] extern crate libfuzzer_sys;
+extern crate png;
+
+fuzz_target!(|data: (u8, u8, u8, u8, Vec<u8>)| {
+    if let Some((raw, encoded)) = encode_png(data.0, data.1, data.2, data.3, &data.4) {
+        let (_info, raw_decoded) = decode_png(&encoded);
+        // raw_decoded can be padded with zeroes at the end, not sure if that's correct
+        let raw_decoded = &raw_decoded[..raw.len()];
+        assert_eq!(raw, raw_decoded);
+    }
+});
+
+fn encode_png(height: u8, filter: u8, compression: u8, color_type: u8, data: &[u8]) -> Option<(&[u8], Vec<u8>)> {
+    // Convert untyped bytes to the correct types and validate them:
+    // height
+    let height = height as u32;
+    if height == 0 { return None }
+    // filter
+    let filter = FilterType::from_u8(filter)?;
+    // color type
+    let color_type = ColorType::from_u8(color_type)?;
+    if let ColorType::Indexed = color_type {
+        return None // TODO: palette needs more data, not supported yet
+    }
+    // compression
+    let compression = match compression {
+        0 => png::Compression::Default,
+        1 => png::Compression::Fast,
+        2 => png::Compression::Best,
+        3 => png::Compression::Huffman,
+        4 => png::Compression::Rle,
+        _ => return None,
+    };
+
+    // infer the rest of the parameters
+    let bytes_per_pixel = color_type.samples() as u32; // safe to cast - actually u8 internally
+    let width = data.len() as u32 / height / bytes_per_pixel;
+    if width == 0 { return None }
+    let total_pixels = height.checked_mul(width)?;
+    let total_bytes = total_pixels.checked_mul(bytes_per_pixel)?;
+
+    // perform the PNG encoding
+    let data_to_encode = &data[..total_bytes as usize];
+    let mut output: Vec<u8> = Vec::new();
+    { // scoped so that we could return the resulting Vec at the end
```

Comment:
Is this scope needed? Now that Rust uses non-lexical lifetimes, I'd expect it to work without a new scope

Reply:
Not anymore. Removed

Reply:
Oh wait this **is** necessary. And now that I've removed it, it doesn't compile :facepalm: 

I thought my IDE ran `cargo check`, but it didn't because fuzzing is a separate folder that's not part of the workspace

Reply:
The CI tests don't catch this either?

Reply:
Fuzzing isn't run on CI and fuzz/ folder is excluded from the workspace because it requires nightly, not to break compilation of the main crate on stable.

oss-fuzz might catch it later, if they run your fuzzers.